### ISSUE TITLE
(feature) <svg> element

### ISSUE THREAD

**Issue by necolas** [facebook/react-strict-dom#4]
### Describe the feature request

Support the [`<svg>` element](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/svg). This element is not currently supported on native.

**Comment by nmn** [comment]
Related: https://www.npmjs.com/package/react-native-svg

**Comment by RyanGuild** [comment]
@necolas and @nmn  I am using rsd for a project and I would like to get involved. Are we ok with a react-native-svg dependency because I think I could whip up the re-export of the rn svg library PR. I believe that stylex already supports svg on web but I would need to look into how rn svg handles styles.

**Comment by necolas** [comment]
Hi, I'm not looking for contributions in this specific area. There is some ongoing refactoring and performance tuning of the existing code before parts of it can be reused for SVG elements. And we have to seamlessly replace an internal stop-gap for cross-platform SVG, which is also being evaluated for replacement with alternative native vector graphics, due to the high cost of rendering react-native-svg elements today.

**Comment by nmn** [comment]
> Is wrapping react-native-svg In a compat.native the way to go? Or maybe using RN Image referencing to bundled assets in the host native app by Id

Both approaches have different trade-offs that are generally app-specific, but both approaches are viable. Use the approach that works best for your app.

**Comment by efoken** [comment]
@axyz I just published a library `react-strict-dom-svg` which exports all common SVG elements as `svg.*` that are also supported by React Native SVG. I already used this approach in some projects, so I thought it's a good idea to publish it as a library, because I was writing the same code again and again 🙂

https://github.com/efoken/react-strict-dom-svg

**Comment by necolas** [comment]
That's not going to be reliable because it's using non-public APIs. For example, `css.props` should not be used on either web or native and shouldn't be expected in future releases. Why don't you continue this SVG stuff to RSD as a PR so it can use those APIs and get updated as they change.

We didn't have svg support because we found the perf of RN SVG was poor. For rendering SVG icons, we ended up using an XML library for Android. I'm not sure if we have support for that on iOS. The engineer who worked on it hasn't contributed it back to RSD. But we should probably figure something out with that too

**Comment by efoken** [comment]
@necolas I thought you are not looking for contributions on SVG 🙃 But of course I can contribute it if you like, would be way better then having another library installed.

I also worked on bigger projects at IBM where we used Vector Drawables for Android, because they were way more performant than RN SVG, but there is no iOS support, so we had to fall back to RN SVG for iOS. We tried many different approaches, but for smaller graphics like icons and stuff, I think there currently is no better/easier alternative, with also app bundle size in mind.

**Comment by axyz** [comment]
@efoken @necolas 

Trying to summarize the current state and the desired future one.

Is it currently possible to consume a component that uses `react-native-svg` from a RSD one? e.g. based on svgr ouputs
```tsx
// exported by an asset library
import Svg, { Path , type SvgProps } from "react-native-svg";
interface SVGRProps { title?: string; titleId?: string; }

export default const MySvgAsset = ({ title, titleId, ...props }: SvgProps & SVGRProps) => (
  <Svg {...} ><Path {...} /></Svg>
);

// icon.ts (some sort of RSD icon wrapper or other component consuming the asset library)
import { MySvgAsset } from "my-assets-library";

<h.div><MySvgAsset /></h.div>
```

Or, instead, an "RSD version" of the library is needed with platform specific files? How would the file differ? Or some other approach I am missing?

Regardint he desired state, would then the goal to be able to get an output from svgr like:
```tsx
// exported by an asset library
import { h as html} from "react-native-svg";
interface SVGRProps { title?: string; titleId?: string; }

export default const MySvgAsset = ({ title, titleId, ...props }: typeof h.svg /* <- likely wrong? */ & SVGRProps) => (
  <h.svg {...} ><h.path {...} /></h.svg>
);

// icon.ts
import { MySvgAsset } from "my-assets-library";

<h.div><MySvgAsset /></h.div>
```

Or is there a scenario where the same `svg/path` web version would work as well?

**Comment by efoken** [comment]
@axyz yes, there should be a scenario, where you have only one SVG component that works on all platforms:

```tsx
import { html } from 'react-strict-dom';

export default const MySvgAsset = (props: React.ComponentProps<typeof html.svg>) => (
  <html.svg {...props}><html.path {...} /></html.svg>
);
```

Besides that, when you use SVGR, there should also be a config option for RSD output 🙂

**Comment by joaqo** [comment]
If anyone wants a more performant version of this, you could use `<img>` on web and `expo-image` on native to load up a svg file. You can wrap it into a simple api of `name`, `size`, and `color` props. Its not as flexible as proper svg support, but may be enough for some use cases.

Also, whats the current status of this? I noticed #272 hasn't been merged after quite a while.

**Comment by MoOx** [comment]
@martinbooth you mentioned in https://github.com/facebook/react-strict-dom/discussions/471#discussioncomment-16896278  _"The performance cost of rendering SVGs has shown to be significant and we avoid them on any of the apps I've worked on."_
I understand that's complex app that Meta are working might want to be careful about this cost, but for a lot of developers, this could be a really cool thing.

It seems that the job done in ##272 would not directly affect RSD performance directly if you don't rely on svg. Do you still think that this should not be in RSD ? Maybe we could do a `react-strict-svg` package like react-strict-animated ? cc @efoken

**Comment by martinbooth** [comment]
> [@martinbooth](https://github.com/martinbooth) you mentioned in [#471 (comment)](https://github.com/facebook/react-strict-dom/discussions/471#discussioncomment-16896278) _"The performance cost of rendering SVGs has shown to be significant and we avoid them on any of the apps I've worked on."_ I understand that's complex app that Meta are working might want to be careful about this cost, but for a lot of developers, this could be a really cool thing.
> 
> It seems that the job done in #[#272](https://github.com/facebook/react-strict-dom/pull/272) would not directly affect RSD performance directly if you don't rely on svg. Do you still think that this should not be in RSD ? Maybe we could do a `react-strict-svg` package like react-strict-animated ? cc [@efoken](https://github.com/efoken)

The problem with adding this to RSD is that we are effectively endorsing that approach and we'd be endorsing it whether your app is sensitive to the performance cost or not.

The app really doesn't have to be complex either to be affected.. if you render a scrollable list containing svgs it will drop frames even on powerful devices and this is not an easy issue to diagnose if it happens

The approach suggested by [axyz](https://github.com/axyz) in https://github.com/facebook/react-strict-dom/issues/4#issuecomment-2663579076 seems like it should work if people really want to use svgs despite these performance concerns and it's probably pretty similar to what was done at meta until the costs were better understood

**Comment by hoangvvo** [comment]
One of my main use case is to be able to set `currentColor` for the svg. Properties of color is usually inherited as well so react-strict-dom might need to export a way to access the internal inherited styles context.

```
export {
    useInheritedStyles as useInheritedStyles_DO_NOT_USE,
    ProvideInheritedStyles as ProvideInheritedStyles_DO_NOT_USE
  } from './modules/ContextInheritedStyles';
```

I am curious how we are using svg at scale at Meta.

**Comment by MoOx** [comment]
OK, so my idea would be to document this choice (no SVG endorsement).

The user should know why this is not a good idea.

- Did you have some benchmark to share?
- What platform was affected because of the performance issue?
- Maybe we could see if that's a problem on a single platform to see if we can improve downstream?
- Is this still an issue with virtualized list ?

I totally understand the rationale behind this SVG choice, but I would like to ensure that we cannot improve/fix this before offering the community another approach (for icons mostly) that can be shared across platforms.
I think having a documented way to say to the user "here is how you can handle your icons" could be a good thing.

**Comment by axyz** [comment]
Thinking beyond just svg, I think that in general would be a good idea to explore a good public API to extend react-strict-dom capabilities.

Many features will be quite hard to be included in the core depending on the limitations in react native itself at a given time. But the community could benefit from an ecosystem of plugins that, while not 100% spec compliant and performant, may be good enough for many use cases, as well as offering a space to iterate and possibly move to core in the future.

Thisican also help with adoption when the current state is not good enough for different reason. For example I could imagine a community polyfill for dialog based on RN modal being a candidate for integration in core in the future, but also another competing one based on bottom-sheet or other widely used libraries that can bridge the gaps in the mid/short term.

Will need to dig deeper on what would be needed, likely at least:
- access to native elements refs
- middleware to modify props and styles
- ways to consume contexts for style inheritance
- babel and postcss to include extensions via config
- probably more...

curious to hear your thoughts.

**Comment by Abbondanzo** [comment]
To provide a bit more context from the Meta side, I had helped develop support for rendering XML vector drawables in React Native to fix the performance and scaling issues we saw from react-native-svg in Android specifically. I added a small blurb in [our blog post in a release from years prior](https://reactnative.dev/blog/2025/02/19/react-native-0.78#performance--quality) explaining this decision and how to wield vector drawables, but I didn't do a good job at communicating this more broadly, my apologies. The issues primarily fell into two buckets:

## Performance
We had (and still do) render non-complex SVGs using react-native-svg in quite a few places. However, they come with a cost--each component has an overhead: in the mounting manager, in Fabric, in the Android view hierarchy as `ReactViewGroup` subclasses, and in iOS as `UIView`. Simple SVGs with one or two elements in them aren't as costly, but more complex shapes with 20, 30, 40 elements can and will cause dropped frames when drawing to your screen.

Take this SVG for example, one of the reaction icons from Meta:

<img width="48" height="48" alt="Image" src="https://github.com/user-attachments/assets/caa09dd3-4b03-495c-aa0a-b9a2c9ae673c" />

It's only comprised of 9 elements inside one group element but in our measurements takes:
- ~3ms to create all the necessary nodes via UIManager (albeit on the JS thread)
- ~7ms to premount each of the views
- ~5ms to draw

Now scale that up to two or three icons per post with two or three posts on screen and you've exhausted your entire frame 6 times over. Granted, this is the more complex version and one can optimize away certain details that are lost as the icon is smaller, but this was the pattern we kept identifying: a handful of dropped frames each time a new batch of these components needed to be put on screen. On average, we saw about 90-100ms wasted each app start rendering these.

When switching over to vector drawables, that cost dropped to just 4ms for 8-10 binary XML inflations. Best of all, XML only needs to be inflated once, in one of Fresco's threads, and the vector drawable is copied nearly instantly wherever it's needed. So if you render the same icons in many places, you aren't paying anywhere close to the same costs as you would with react-native-svg. Lastly, it's a single image element with a single set of props, single set of event handlers, and single layout pass.

While we placed heavy emphasis on Android performance and issues, not as much attention was paid to iOS. I dabbled with svg2pdf conversion and universally scaling icons but neither were as quick, performant, or intuitive as just carrying around the classic `1x`, `2x`, `3x` PNGs for icons. Designers really care for that pixel-perfect rasterization and scalable vector graphics break all assumptions.

## Scaling

This has been documented in [a few spots](https://github.com/software-mansion/react-native-svg/issues/1451) and some fixes have since landed, but at the time react-native-svg struggled with `transform` scaling that would cause aliasing artifacts. We found that vector drawables could comfortably handle this out of the box so we leaned towards vector drawables.

# Recommendations

This is my own personal recommendation, not as a maintainer of RSD but seeing what has worked across other RN libraries, but react-native-svg should probably be adapted and offered through a new, supported package in this project rather than being added to react-strict-dom. Otherwise, it runs the risk of inflating projects which do not rely on the library and instead rely on others like react-native-vector-icons. Functionally, the components offered can be the same compat.native wrappers that everyone would already be using. 

The react-native-svg library is perfectly fine for most use cases. It's also way more powerful for non-static icons that might need to change shape or have attributes passed through them--something that neither vector drawables nor rasterized PNGs can solve. But it's not needed _everywhere_ and that's okay too.

**Comment by martinbooth** [comment]
Thanks @Abbondanzo for the very comprehensive explanation of why we avoid SVGs with react native at meta.

I think the theme here of wanting to be able to extend RSD when features don't necessarily align with RSD's goals, or are too niche to include in the main RSD package is important so I think next steps would be to figure out what is missing so that we can build this (and potentially other features) in a separate package that can be optionally installed by those who need the functionality

We can either make the necessary apis available so that svg functionality can live outside of the RSD repo and not rely on private implementation details.. for example the https://github.com/efoken/react-strict-dom-svg approach

Or, we can put the functionality in this repo alongside the main RSD package.. for example under packages/react-strict-svg

Given this functionality isn't really something meta would want for the reasons provided above, it's probably not something I'll be able to get to that soon, but if anyone does want to take it on, I'll make sure either of those approaches gets merged (or even better, both)

@hoangvvo I don't think currentColor is supported by RN currently, but this would be a useful feature to have

@MoOx There are other options beside svg.. there's the approach Peter mentioned, and things like https://github.com/oblador/react-native-vector-icons are also options, which is why I'd prefer to keep svg optional

@axyz These are good suggestions and probably some of them are required to implement svg as an optional package

**Comment by hoangvvo** [comment]
> I don't think currentColor is supported by RN currently, but this would be a useful feature to have

react-native-svg supports `currentColor` via the color properties on the Svg or the `color` style props. (However, their priority is reversed compared to the web version -- that is, on the web, if `color` is defined in styles/css, it overrides the `color` prop set on the `svg` element, but this is not the case for react-native-svg). Docs: https://github.com/software-mansion/react-native-svg/blob/main/USAGE.md

```
<Svg color="#fff" />
// or
<Svg style={{ color: "#fff" }} />
```

I did look at using the `compat.native` way and find that it is possible to work with something like below:

```tsx
import { compat } from "react-strict-dom";
import NativeSvg, { Path } from "react-native-svg";

function Svg({ style, children, ...props }) {
  return (
    <compat.native as="span" style={style}>
      {(nativeProps) => (
        <NativeSvg {...props} {...nativeProps}>
          {children}
        </NativeSvg>
      )}
    </compat.native>
  );
}

function DemoBoltIcon({ style, ...rest }) {
  const props = {
    ...rest,
    ...css.props(style),
  };

  return (
    <Svg viewBox="0 0 24 24" fill="none" {...props}>
      <Path
        d="M13 2L6 13H11L10 22L18 10H13L13 2Z"
        stroke="currentColor"
        strokeLinecap="round"
        strokeLinejoin="round"
      />
    </Svg>
  );
}
```

In this case `nativeProps.style` will have the `color` set based on its parent. This works, but it feels much comfortable with the internal access to the inherited styles instead of using the `span` compat when an svg is not one.

In my workflow, I write a library, which allows me to use `@svgr/core` to generate the svg for both web and react-native. 

On the web, it simply outputs something like:

```tsx
function DemoBoltIcon({ style, ...rest }) {
  const props = {
    ...rest,
    ...css.props(style),
  };

  return (
    <svg viewBox="0 0 24 24" fill="none" {...props}>
      <path
        d="M13 2L6 13H11L10 22L18 10H13L13 2Z"
        stroke="currentColor"
        strokeLinecap="round"
        strokeLinejoin="round"
      />
    </svg>
  );
}
```

Using a transformer would be cleaner compared to implementing every single svg nodes as the approach in https://github.com/efoken/react-strict-dom-svg, if we only need to touch the outer svg tag and not those inside.

**Comment by MoOx** [comment]
I know [react-native-vector-icons](https://github.com/oblador/react-native-vector-icons) but I usually use my own svg for projects, using [react-from-svg](https://github.com/MoOx/react-from-svg).
I found it way easier to use and work on svg rather than having to maintain a font. I totally understand the issue with lots of svg, but for some icons, when there are just a few, it's probably acceptable. But I get the idea to not get that in core of RSD.

We should find a way to support it tho. Maybe a documented external package could be a thing. Would you be ok if we add documentation about this (recommendation on icons/svg when to / when not to + link to @efoken package) ?

**Comment by martinbooth** [comment]
I'm wondering if just exposing the currently internal useInheritedTextStyle as an unstable api is enough to unblock svg support as per https://github.com/facebook/react-strict-dom/commit/70d57bf727dd44be85faaa39c4b905f91c57307b

I would be happy to merge @efoken's PR (https://github.com/facebook/react-strict-dom/pull/272) if the functionality was added as a sibling package, similar to how [react-strict-animated](https://github.com/facebook/react-strict-dom/tree/main/packages/react-strict-animated) was done. That way I can keep the code working if there is a need to change the internal useInheritedTextStyle api.

While using compat.native and as="span" does expose the inherited color, it likely has unintended side effects such as disabling block emulation, telling child components that they are text descendants when they're not, etc

I think that with useInheritedTextStyle, compat.native can be used to implement the svg element correctly:

For example, a helper function like this can render the component and fetch the inherited color using the hook:

```
function createSvgComponent(
  NativeComponent,
  displayName
) {
  function Inner({
    nativeProps,
    children,
    ...svgProps
  }) {
    const inherited = useInheritedTextStyle_UNSTABLE();
    const color = svgProps.color ?? inherited.color;
    return (
      <NativeComponent
        {...nativeProps}
        {...svgProps}
        color={color}
      >
        {children}
      </NativeComponent>
    );
  }
  Inner.displayName = `${displayName}.Inner`;

  function SvgComponent(props) {
    const { style, children, ...svgProps } = props;
    return (
      <compat.native style={style}>
        {(nativeProps: AnyProps) => (
          <Inner nativeProps={nativeProps} {...svgProps}>
            {children}
          </Inner>
        )}
      </compat.native>
    );
  }
  SvgComponent.displayName = displayName;
  return SvgComponent;
}
```

It is maybe not the most ergonomic code requiring that inner component, but keen to get some feedback and find out if there's anything else required internally from RSD besides the inherited color.

I am also happy to put together an implementation if required, but given how close @efoken's PR is, I'd like to make sure the right people get the credit for doing this so would rather help get the work already done merged if possible

**Comment by axyz** [comment]
Was also exploring this option: https://github.com/facebook/react-strict-dom/discussions/479#discussioncomment-17076710

beside `html.*` the core could also export an unimplemented `svg.*` namespace to be imported and overridden optionally 

While the rest of the missing links could be handled indeed with the improvements to compat.native and additional hooks

That should in theory allow consumers to use the react-native-svg version or, where needed, explore a version made using vector-icons (even if it only covers a small subset of specs)
